maven-model-builder/src/main/java/org/apache/maven/model/io/DefaultModelReader.java

```diff
@@ -27,21 +27,26 @@
 import java.util.Map;
 import java.util.Objects;
 
+import javax.inject.Named;
+import javax.inject.Singleton;
+
 import org.apache.maven.model.InputSource;
 import org.apache.maven.model.Model;
 import org.apache.maven.model.io.xpp3.MavenXpp3Reader;
 import org.apache.maven.model.io.xpp3.MavenXpp3ReaderEx;
-import org.codehaus.plexus.component.annotations.Component;
 import org.codehaus.plexus.util.ReaderFactory;
 import org.codehaus.plexus.util.xml.XmlStreamReader;
 import org.codehaus.plexus.util.xml.pull.XmlPullParserException;
+import org.eclipse.sisu.Typed;
 
 /**
  * Handles deserialization of a model from some kind of textual format like XML.
  *
  * @author Benjamin Bentmann
  */
-@Component( role = ModelReader.class )
+@Named
+@Singleton
+@Typed( ModelReader.class )
```

Comment:
Why is the `@Typed` here?

Reply:
This one case is ambiguous; without it, `org.apache.maven.repository.internal.RepositorySystemTest` fails with `StackOverflow` when `DefaultModelProcessor` gets itself injected as its own `ModelReader` (because `ModelProcessor` extends `ModelReader`).

There may be a better approach, but that's why it's here.

Reply:
I understand the fact about the recursion, but I don't understand why *this* class requires it. Isn't the type here straight forward? `DefaultModelProcessor` is aready annotated with `@Component( role = ModelProcessor.class )` how can it still be injected as `ModelReader`?

Reply:
You're right, this is just repeating the (obvious) type that would be inferred. Testing again, it doesn't seem to be necessary.

However, on `DefaultModelProcessor`, with `Component` now replaced by `@Named @Singleton`, it's made available as *all* its implemented interfaces. I've added a `@Typed` to restrict that to `ModelProcessor`.

I'm not able to explain why I no longer see recursive wiring, but that `@Typed` should prevent it.